File: dotnet/samples/04-DynamicRag/Plugins/MathPlugin/Math.cs
// Copyright (c) Microsoft. All rights reserved.

using System.ComponentModel;
using Microsoft.SemanticKernel;
using Microsoft.SemanticKernel.Handlebars;


public class Math
{

    public Math()
    {
    }

    [SKFunction]
    [Description("Uses functions from the Math plugin to solve math problems.")]
    [SKOutputDescription("The answer to the math problem.")]
    [SKSample(
        inputs: "{\"math_problem\",\"If I started with $120 in the stock market, how much would I have after 10 years if the growth rate was 5%?\"}",
        output:"After 10 years, starting with $120, and with a growth rate of 5%, you would have $195.47 in the stock market."
    )]
    public static async Task<string> PerformMath(
        IKernel kernel,
        [Description("A description of a math problem; use the GenerateMathProblem function to create one.")] string math_problem
    )
    {
        int maxTries = 1;
        HandlebarsPlan? lastPlan = null;
        Exception? lastError = null;

        while (maxTries >= 0)
        {
            // Create the planner
            var planner = new HandlebarsPlanner(kernel, new HandlebarsPlannerConfiguration(){
                IncludedPlugins = new () { "Math" },
                ExcludedFunctions = new () { "Math.PerformMath", "Math.GenerateMathProblem" },
                LastPlan = lastPlan, // Pass in the last plan in case we want to try again
                LastError = lastError?.Message // Pass in the last error to avoid trying the same thing again
            });
            var plan = await planner.CreatePlanAsync("Solve the following math problem.\n\n" + math_problem);
            lastPlan = plan;


            // Console.ForegroundColor = ConsoleColor.Blue;
            // Console.WriteLine("[Plan]");
            // Console.ForegroundColor = ConsoleColor.DarkBlue;
            // Console.WriteLine(lastPlan.ToString().Trim());
            // Console.ResetColor();
            
            // Run the plan
            try {
                var result = await plan.InvokeAsync(kernel, new Dictionary<string, object?>());

                // Console.ForegroundColor = ConsoleColor.Green;
                // Console.WriteLine("[Result]");
                // Console.ForegroundColor = ConsoleColor.DarkGreen;
                // Console.WriteLine(result.ToString().Trim());
                // Console.ResetColor();

                return result.GetValue<string>()!;
            } catch (Exception e) {
                // If we get an error, try again
                lastError = e;

                // Console.ForegroundColor = ConsoleColor.Red;
                // Console.WriteLine("[Error]");
                // Console.ForegroundColor = ConsoleColor.DarkRed;
                // Console.WriteLine(e.Message.ToString().Trim());
                // Console.ResetColor();
            }
            maxTries--;
        }

        // If we tried too many times, throw an exception
        throw lastError!;
    }
    

    [SKFunction]
    [Description("Adds two numbers.")]
    [SKOutputDescription("The summation of the numbers.")]
    [SKSample(
        inputs: "{\"number1\":1, \"number2\":2}",
        output:"3"
    )]
    public static double Add(
        [Description("The first number to add")] double number1,
        [Description("The second number to add")] double number2
    )
    {
        return number1 + number2;
    }

    [SKFunction]
    [Description("Subtracts two numbers.")]
    [SKOutputDescription("The difference between the minuend and subtrahend.")]
    [SKSample(
        inputs: "{\"number1\":5, \"number2\":2}",
        output:"3"
    )]
    public static double Subtract(
        [Description("The minuend")] double number1,
        [Description("The subtrahend")] double number2
    )
    {
        return number1 - number2;
    }

    [SKFunction]
    [Description("Multiplies two numbers.")]
    [SKOutputDescription("The product of the numbers.")]
    [SKSample(
        inputs: "{\"number1\":5, \"number2\":2}",
        output:"10"
    )]
    public static double Multiply(
        [Description("The first number to multiply")] double number1,
        [Description("The second number to multiply")] double number2
    )
    {
        return number1 * number2;
    }

    [SKFunction]
    [Description("Divides two numbers.")]
    [SKOutputDescription("The quotient of the dividend and divisor.")]
    [SKSample(
        inputs: "{\"number1\":10, \"number2\":2}",
        output:"5"
    )]
    public static double Divide(
        [Description("The dividend")] double number1,
        [Description("The divisor")] double number2
    )
    {
        return number1 / number2;
    }

    [SKFunction]
    [Description("Gets the remainder of two numbers.")]
    [SKOutputDescription("The remainder of the dividend and divisor.")]
    [SKSample(
        inputs: "{\"number1\":10, \"number2\":3}",
        output:"1"
    )]
    public static double Modulo(
        [Description("The dividend")] double number1,
        [Description("The divisor")] double number2
    )
    {
        return number1 % number2;
    }

    [SKFunction]
    [Description("Gets the absolute value of a number.")]
    [SKOutputDescription("The absolute value of the number.")]
    [SKSample(
        inputs: "{\"number1\":-10}",
        output:"5"
    )]
    public static double Abs(
        [Description("The number")] double number1
    )
    {
        return System.Math.Abs(number1);
    }

    [SKFunction]
    [Description("Gets the ceiling of a single number.")]
    [SKOutputDescription("The ceiling of the number.")]
    [SKSample(
        inputs: "{\"number1\":5.1}",
        output:"6"
    )]
    public static double Ceil(
        [Description("The number")] double number1
    )
    {
        return System.Math.Ceiling(number1);
    }

    [SKFunction]
    [Description("Gets the floor of a single number.")]
    [SKOutputDescription("The floor of the number.")]
    [SKSample(
        inputs: "{\"number1\":5.9}",
        output:"5"
    )]
    public static double Floor(
        [Description("The number")] double number1
    )
    {
        return System.Math.Floor(number1);
    }

    [SKFunction]
    [Description("Gets the maximum of two numbers.")]
    [SKOutputDescription("The maximum of the two numbers.")]
    [SKSample(
        inputs: "{\"number1\":5, \"number2\":10}",
        output:"10"
    )]
    public static double Max(
        [Description("The first number")] double number1,
        [Description("The second number")] double number2
    )
    {
        return System.Math.Max(number1, number2);
    }

    [SKFunction]
    [Description("Gets the minimum of two numbers.")]
    [SKOutputDescription("The minimum of the two numbers.")]
    [SKSample(
        inputs: "{\"number1\":5, \"number2\":10}",
        output:"5"
    )]
    public static double Min(
        [Description("The first number")] double number1,
        [Description("The second number")] double number2
    )
    {
        return System.Math.Min(number1, number2);
    }

    [SKFunction]
    [Description("Gets the sign of a number.")]
    [SKOutputDescription("The sign of the number.")]
    [SKSample(
        inputs: "{\"number1\":-10}",
        output:"-1"
    )]
    public static double Sign(
        [Description("The number")] double number1
    )
    {
        return System.Math.Sign(number1);
    }

    [SKFunction]
    [Description("Gets the square root of a number.")]
    [SKOutputDescription("The square root of the number.")]
    [SKSample(
        inputs: "{\"number1\":25}",
        output:"5"
    )]
    public static double Sqrt(
        [Description("The number")] double number1
    )
    {
        return System.Math.Sqrt(number1);
    }

    [SKFunction]
    [Description("Gets the sine of a number.")]
    [SKOutputDescription("The sine of the number.")]
    [SKSample(
        inputs: "{\"number1\":0}",
        output:"0"
    )]
    public static double Sin(
        [Description("The number")] double number1
    )
    {
        return System.Math.Sin(number1);
    }

    [SKFunction]
    [Description("Gets the cosine of a number.")]
    [SKOutputDescription("The cosine of the number.")]
    [SKSample(
        inputs: "{\"number1\":0}",
        output:"1"
    )]
    public static double Cos(
        [Description("The number")] double number1
    )
    {
        return System.Math.Cos(number1);
    }

    [SKFunction]
    [Description("Gets the tangent of a number.")]
    [SKOutputDescription("The tangent of the number.")]
    [SKSample(
        inputs: "{\"number1\":0}",
        output:"0"
    )]
    public static double Tan(
        [Description("The number")] double number1
    )
    {
        return System.Math.Tan(number1);
    }

    [SKFunction]
    [Description("Raises a number to a power.")]
    [SKOutputDescription("The number raised to the power.")]
    [SKSample(
        inputs: "{\"number1\":5, \"number2\":2}",
        output:"25"
    )]
    public static double Pow(
        [Description("The number")] double number1,
        [Description("The power")] double number2
    )
    {
        return System.Math.Pow(number1, number2);
    }

    [SKFunction]
    [Description("Gets the natural logarithm of a number.")]
    [SKOutputDescription("The natural logarithm of the number.")]
    [SKSample(
        inputs: "{\"number1\":10, \"number2\":10}",
        output:"2.302585092994046"
    )]
    public static double Log(
        [Description("The number")] double number1,
        [Description("The base of the logarithm")] double number2 = 10
    )
    {
        return System.Math.Log(number1, number2);
    }

    [SKFunction]
    [Description("Gets a rounded number.")]
    [SKOutputDescription("The rounded number.")]
    [SKSample(
        inputs: "{\"number\":1.23456, \"digits\":2}",
        output:"1.23"
    )]
    public static double Round(
        [Description("The number")] double number1,
        [Description("The number of digits to round to")] int number2 = 0
    )
    {
        return System.Math.Round(number1, number2);
    }
}

File: python/samples/04-DynamicRag/Plugins/MathPlugin/math.py
import math
from enum import Enum
from typing import TYPE_CHECKING

from python.src.planners.handlebars_planner import (
    HandleBarsPlanner,
    HandleBarsPlannerConfig,
)
from python.src.plugins import (
    sk_function,
    sk_function_parameter,
)

if TYPE_CHECKING:
    from python.src.kernel import NewKernel


class console_colors(str, Enum):
    OKBLUE = "\033[94m"
    OKGREEN = "\033[92m"
    FAIL = "\033[91m"
    ENDC = "\033[0m"


class Math:
    def __init__(self):
        self.kernel = None

    @sk_function(
        description="Uses functions from the Math plugin to solve math problems.",
        name="PerformMath",
    )
    @sk_function_parameter(
        name="kernel",
        description="The kernel to use for planning.",
        required=True,
        type="kernel",
    )
    @sk_function_parameter(
        name="math_problem",
        description="A description of a math problem; use the GenerateMathProblem function to create one.",
        required=True,
        type="string",
    )
    @sk_function_parameter(
        name="result",
        direction="output",
        description="The answer to the math problem.",
        type="number",
    )
    async def perform_math(self, variables, **kwargs):
        kernel: "NewKernel" = kwargs.get("kernel")
        assert kernel
        math_problem: str = variables["math_problem"]
        max_tries = 1
        last_plan = None
        last_error = None
        # TODO: new config that also has include plugin
        while max_tries >= 0:
            config = HandleBarsPlannerConfig(
                excluded_plugins=["Intent"],
                excluded_functions=[
                    "Math_GenerateMathProblem",
                    "Math_PerformMath",
                ],
                last_plan=last_plan,
                last_error=last_error,
            )
            planner = HandleBarsPlanner(kernel=kernel, configuration=config)
            plan = await planner.create_plan(
                goal=f"Solve the following math problem: {math_problem}"
            )
            last_plan = str(plan)
            print(f"{console_colors.OKBLUE}[Plan]: {plan}\n{console_colors.ENDC}")
            try:
                result = await kernel.run_async(plan, variables, **kwargs)
                # result = await plan.run_async(variables)
                print(
                    f"{console_colors.OKGREEN}[Result]: {result['result']}\n{console_colors.ENDC}"
                )
                return str(result["result"])
            except Exception as exc:
                print(f"{console_colors.FAIL}[Error]: {exc}\n{console_colors.ENDC}")
                last_error = str(exc)
                max_tries -= 1

    def _get_variables(self, args, kwargs, function):
        if "variables" in kwargs and kwargs["variables"]:
            return kwargs["variables"]
        variables = {}
        idx = 0
        for parameter in function.__sk_function_context_parameters__:
            if parameter["direction"] == "output":
                continue
            if parameter["name"] in kwargs:
                variables[parameter["name"]] = kwargs[parameter["name"]]
            elif args:
                variables[parameter["name"]] = args[idx]
            if (
                variables[parameter["name"]] is None
                and parameter["default_value"] is not None
            ):
                variables[parameter["name"]] = parameter["default_value"]
            idx += 1
        return variables

    @sk_function(
        description="Adds two numbers",
        name="Add",
    )
    @sk_function_parameter(
        name="number1",
        description="The first number to add",
        required=True,
        type="number",
    )
    @sk_function_parameter(
        name="number2",
        description="The second number to add",
        required=True,
        type="number",
    )
    @sk_function_parameter(
        name="sum",
        direction="output",
        description="The addition result",
        type="number",
        required=True,
    )
    def add(self, *args, **kwargs):
        variables = self._get_variables(args, kwargs, self.add)
        if "number1" not in variables or "number2" not in variables:
            raise ValueError("Missing number1 or number2 in both args and kwargs.")
        return float(variables["number1"]) + float(variables["number2"])

    @sk_function(
        description="Subtracts two numbers",
        name="Subtract",
    )
    @sk_function_parameter(
        name="number1",
        description="The minuend",
        required=True,
        type="number",
    )
    @sk_function_parameter(
        name="number2",
        description="The subtrahend",
        required=True,
        type="number",
    )
    @sk_function_parameter(
        name="difference",
        direction="output",
        description="The difference between the minuend and subtrahend.",
        type="number",
        required=True,
    )
    def subtract(self, *args, **kwargs):
        variables = self._get_variables(args, kwargs, self.subtract)
        if "number1" not in variables or "number2" not in variables:
            raise ValueError("Missing number1 or number2 in both args and kwargs.")
        return float(variables["number1"]) - float(variables["number2"])

    @sk_function(
        description="Multiplies two numbers",
        name="Multiply",
    )
    @sk_function_parameter(
        name="number1",
        description="The first number to multiply",
        required=True,
        type="number",
    )
    @sk_function_parameter(
        name="number2",
        description="The second number to multiply",
        required=True,
        type="number",
    )
    @sk_function_parameter(
        name="product",
        direction="output",
        description="The product of the numbers.",
        type="number",
        required=True,
    )
    def multiply(self, *args, **kwargs):
        variables = self._get_variables(args, kwargs, self.multiply)
        if "number1" not in variables or "number2" not in variables:
            raise ValueError("Missing number1 or number2 in both args and kwargs.")
        return float(variables["number1"]) * float(variables["number2"])

    @sk_function(
        description="Divides two numbers",
        name="Divide",
    )
    @sk_function_parameter(
        name="number1",
        description="The dividend",
        required=True,
        type="number",
    )
    @sk_function_parameter(
        name="number2",
        description="The divisor",
        required=True,
        type="number",
    )
    @sk_function_parameter(
        name="quotient",
        direction="output",
        description="The quotient of the dividend and divisor.",
        type="number",
        required=True,
    )
    def divide(self, *args, **kwargs):
        variables = self._get_variables(args, kwargs, self.divide)
        if "number1" not in variables or "number2" not in variables:
            raise ValueError("Missing number1 or number2 in both args and kwargs.")
        return float(variables["number1"]) / float(variables["number2"])

    @sk_function(
        description="Finds the remainder of two numbers",
        name="Modulo",
    )
    @sk_function_parameter(
        name="number1",
        description="The dividend",
        required=True,
        type="number",
    )
    @sk_function_parameter(
        name="number2",
        description="The divisor",
        required=True,
        type="number",
    )
    @sk_function_parameter(
        name="remainder",
        direction="output",
        description="The remainder of the dividend and divisor.",
        type="number",
        required=True,
    )
    def modulo(self, *args, **kwargs):
        variables = self._get_variables(args, kwargs, self.modulo)
        if "number1" not in variables or "number2" not in variables:
            raise ValueError("Missing number1 or number2 in both args and kwargs.")
        return float(variables["number1"]) % float(variables["number2"])

    @sk_function(
        description="Gets the absolute value of a number",
        name="Absolute",
    )
    @sk_function_parameter(
        name="number",
        description="The number to get the absolute value of",
        required=True,
        type="number",
    )
    @sk_function_parameter(
        name="absolute",
        direction="output",
        description="The absolute value of the number.",
        type="number",
        required=True,
    )
    def absolute(self, *args, **kwargs):
        variables = self._get_variables(args, kwargs, self.absolute)
        if "number" not in variables:
            raise ValueError("Missing number in both args and kwargs")
        return abs(float(variables["number"]))

    @sk_function(name="Ceil", description="Gets the ceiling of a single number.")
    @sk_function_parameter(
        name="number",
        description="The number to get the ceiling of.",
        required=True,
        type="number",
    )
    @sk_function_parameter(
        name="ceiling",
        direction="output",
        description="The ceiling of the number.",
        type="number",
        required=True,
    )
    def ceil(self, *args, **kwargs):
        variables = self._get_variables(args, kwargs, self.ceil)
        if "number" not in variables:
            raise ValueError("Missing number in both args and kwargs")
        return math.ceil(float(variables["number"]))

    @sk_function(name="Floor", description="Gets the floor of a single number.")
    @sk_function_parameter(
        name="number",
        description="The number to get the floor of.",
        required=True,
        type="number",
    )
    @sk_function_parameter(
        name="floor",
        direction="output",
        description="The floor of the number.",
        type="number",
        required=True,
    )
    def floor(self, *args, **kwargs):
        variables = self._get_variables(args, kwargs, self.floor)
        if "number" not in variables:
            raise ValueError("Missing number in both args and kwargs")
        return math.floor(float(variables["number"]))

    @sk_function(
        name="Max",
        description="Gets the maximum value of two numbers.",
    )
    @sk_function_parameter(
        name="number1",
        description="The first number to compare.",
        required=True,
        type="number",
    )
    @sk_function_parameter(
        name="number2",
        description="The second number to compare.",
        required=True,
        type="number",
    )
    @sk_function_parameter(
        name="max",
        direction="output",
        description="The maximum value of the two numbers.",
        type="number",
        required=True,
    )
    def max(self, *args, **kwargs):
        variables = self._get_variables(args, kwargs, self.max)
        if "number1" not in variables or "number2" not in variables:
            raise ValueError("Missing number1 or number2 in both args and kwargs.")
        return max(float(variables["number1"]), float(variables["number2"]))

    @sk_function(
        name="Min",
        description="Gets the minimum value of two numbers.",
    )
    @sk_function_parameter(
        name="number1",
        description="The first number to compare.",
        required=True,
        type="number",
    )
    @sk_function_parameter(
        name="number2",
        description="The second number to compare.",
        required=True,
        type="number",
    )
    @sk_function_parameter(
        name="min",
        direction="output",
        description="The minimum value of the two numbers.",
        type="number",
        required=True,
    )
    def min(self, *args, **kwargs):
        variables = self._get_variables(args, kwargs, self.min)
        if "number1" not in variables or "number2" not in variables:
            raise ValueError("Missing number1 or number2 in both args and kwargs.")
        return min(float(variables["number1"]), float(variables["number2"]))

    @sk_function(
        name="Sign",
        description="Gets the sign of a number.",
    )
    @sk_function_parameter(
        name="number",
        description="The number to get the sign of.",
        required=True,
        type="number",
    )
    @sk_function_parameter(
        name="sign",
        direction="output",
        description="The sign of the number.",
        type="number",
        required=True,
    )
    def sign(self, *args, **kwargs):
        variables = self._get_variables(args, kwargs, self.sign)
        if "number" not in variables:
            raise ValueError("Missing number in both args and kwargs")
        return math.copysign(1.0, float(variables["number"]))

    @sk_function(
        name="Sqrt",
        description="Gets the square root of a number.",
    )
    @sk_function_parameter(
        name="number",
        description="The number to get the square root of.",
        required=True,
        type="number",
    )
    @sk_function_parameter(
        name="sqrt",
        direction="output",
        description="The square root of the number.",
        type="number",
        required=True,
    )
    def sqrt(self, *args, **kwargs):
        variables = self._get_variables(args, kwargs, self.sqrt)
        if "number" not in variables:
            raise ValueError("Missing number in both args and kwargs")
        return math.sqrt(float(variables["number"]))

    @sk_function(
        name="Sin",
        description="Gets the sine of a number.",
    )
    @sk_function_parameter(
        name="number",
        description="The number to get the sine of.",
        required=True,
        type="number",
    )
    @sk_function_parameter(
        name="sin",
        direction="output",
        description="The sine of the number.",
        type="number",
        required=True,
    )
    def sin(self, *args, **kwargs):
        variables = self._get_variables(args, kwargs, self.sin)
        if "number" not in variables:
            raise ValueError("Missing number in both args and kwargs")
        return math.sin(float(variables["number"]))

    @sk_function(
        name="Cos",
        description="Gets the cosine of a number.",
    )
    @sk_function_parameter(
        name="number",
        description="The number to get the cosine of.",
        required=True,
        type="number",
    )
    @sk_function_parameter(
        name="cos",
        direction="output",
        description="The cosine of the number.",
        type="number",
        required=True,
    )
    def cos(self, *args, **kwargs):
        variables = self._get_variables(args, kwargs, self.cos)
        if "number" not in variables:
            raise ValueError("Missing number in both args and kwargs")
        return math.cos(float(variables["number"]))

    @sk_function(
        name="Tan",
        description="Gets the tangent of a number.",
    )
    @sk_function_parameter(
        name="number",
        description="The number to get the tangent of.",
        required=True,
        type="number",
    )
    @sk_function_parameter(
        name="tan",
        direction="output",
        description="The tangent of the number.",
        type="number",
        required=True,
    )
    def tan(self, *args, **kwargs):
        variables = self._get_variables(args, kwargs, self.tan)
        if "number" not in variables:
            raise ValueError("Missing number in both args and kwargs")
        return math.tan(float(variables["number"]))

    @sk_function(
        name="Pow",
        description="Gets the power of a number.",
    )
    @sk_function_parameter(
        name="number",
        description="The number to get the power of.",
        required=True,
        type="number",
    )
    @sk_function_parameter(
        name="power",
        description="The power of the number.",
        type="number",
        required=True,
    )
    @sk_function_parameter(
        name="pow",
        direction="output",
        description="The power of the number.",
        type="number",
        required=True,
    )
    def pow(self, *args, **kwargs):
        variables = self._get_variables(args, kwargs, self.pow)
        if "number" not in variables or "power" not in variables:
            raise ValueError("Missing number or power in both args and kwargs")
        return math.pow(float(variables["number"]), float(variables["power"]))

    @sk_function(
        name="Log",
        description="Gets the logarithm of a number.",
    )
    @sk_function_parameter(
        name="number1",
        description="The number to get the logarithm of.",
        required=True,
        type="number",
    )
    @sk_function_parameter(
        name="number2",
        description="The base of the logarithm.",
        type="number",
        default_value=10,
    )
    @sk_function_parameter(
        name="log",
        direction="output",
        description="The logarithm of the number.",
        type="number",
        required=True,
    )
    def log(self, *args, **kwargs):
        variables = self._get_variables(args, kwargs, self.log)
        if "number1" not in variables and "number2" not in variables:
            raise ValueError("Missing number1 in both args and kwargs")
        return math.log(float(variables["number1"]), float(variables["number2"]))

    @sk_function(
        name="Round",
        description="Gets the rounded value of a number.",
    )
    @sk_function_parameter(
        name="number1",
        description="The number to get the rounded value of.",
        required=True,
        type="number",
    )
    @sk_function_parameter(
        name="number2",
        description="The number of digits to round to.",
        required=True,
        type="number",
    )
    @sk_function_parameter(
        name="round",
        direction="output",
        description="The rounded value of the number.",
        type="number",
        required=True,
    )
    def round(self, *args, **kwargs):
        variables = self._get_variables(args, kwargs, self.round)
        if "number1" not in variables or "number2" not in variables:
            raise ValueError("Missing number1 or number2 in both args and kwargs.")
        return round(float(variables["number1"]), int(variables["number2"]))
